win32ss/user/ntuser/msgqueue.c

```diff
@@ -136,19 +137,22 @@ UserSetCursor(
         {
             /* Call GDI to set the new screen cursor */
             PCURICON_OBJECT CursorFrame = NewCursor;
-            if(NewCursor->CURSORF_flags & CURSORF_ACON)
+            if (NewCursor->CURSORF_flags & CURSORF_ACON)
             {
-                FIXME("Should animate the cursor, using only the first frame now.\n");
-                CursorFrame = ((PACON)NewCursor)->aspcur[0];
+                gAniCursor = (PACON)NewCursor;
+                IntSetTimer(pWnd, ID_EVENT_SYSTIMER_ANIMATECURSOR, ((PACON)NewCursor)->ajifRate[((PACON)NewCursor)->iicur] * 100 / 6, SystemTimerProc, TMRF_SYSTEM);
```

Comment:
`IntKillTimer`?

Reply:
As you can see, I'm calling the system timer here (`TMRF_SYSTEM`), which calls `SystemTimerProc` from here, with a new `idEvent` case implemented there. `IntKillTimer` is already called from that function in case when the execution breaks (hence, to avoid breaking when it's not needed, I need to return instead). I've also added the condition to break in my newly implemented `ID_EVNT_SYSTIMER_ANIMATECURSOR` case: it breaks and calls `IntKillTimer` on any invalid member of the following: `ACON` (which is filled by `gAniCursor` set here, `ACON.aspcur`, `ACON.aicur`, `ACON.iicur`, `ACON.aspcur[ACON.aicur[ACON.iicur]]`.
To break the timer on cursor animation when the ani cursor is unset, same as in ant other required case, I simply need to add one more break condition in the appropriate case of `SystemTimerProc`. In that case, `IntKillTimer` will be called automatically. See this and the code above it for details:
https://github.com/oleg-dubinskiy/reactos/blob/2922a840b6bcbc9711e9611b9f33c09566dca7bf/win32ss/user/ntuser/timer.c#L387
I doubt I need to unnecessary call it 2nd time here...

Reply:
OK, but which condition is better to kill the timer? Do we have any function(s) to check is the timer already set? Or it's better to check for a cursor validity instead? :thinking:

Reply:
No condition to be checked. Just kill it.

Reply:
Additionally, you could kill the timer in the timer procedure (Sorry, I'm not sure).

Reply:
Additionally, you could kill the timer at the prologue of the timer procedure (not the window procedures).

Reply:
You mean in `SystemTimerProc`'s case? Then the animation will not work since the timer is immediately killed then I guess.

Reply:
Oh, I see. My mistake. Never mind.

Reply:
(I'm not sure. The below text is just my imagination)
The interval of animation frame is not a constant. When the timer was ignited, then kill the timer and step the frame and set another timer with new interval.

Reply:
> No condition to be checked. Just kill it. Even if you failed to kill it, you are safe from the authority.

Since `UserSetCursor` is called repeatedly from `SystemTimerProc`, the `IntKillTimer` call you previously suggested has no any effect because it isn't called in case `CURSORF_ACON` flag isn't set (the condition is checked). That's why it didn't fix too fas animation issue. That flag is not set by us for each frame, since each of the frames is cosidered as a normal (non-animated) cursor. But there is set `CURSORF_ACONFRAME` flag instead. So perhaps it makes sense to check for it instead.

Reply:
By the way, what is the magic number `6`?

Reply:
> By the way, what is the magic number `6`?

I'm not sure exactly which constant it might be (but definitely it's not a number of system icons, because we're handling cursors here). I even think it's not a constant at all. This is based on a results of my debugging analysis. First I compared the values found via WinDbg with the ones printed in ReactOS. Then I substituted the random numbers to make the values similar to Windows's. They are different for each ANI cursor, but the way how they are calculated is always same. `PACON.ajifRate[PACON.iicur]` itself usually has too small value, to be as a time value (usually it equals `2` or `3`, so it should be recalculated).

Reply:
Here are some examples of how the proper value should look like:
```
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[10] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[11] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[7] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[15] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[0] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[1] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[1] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[2] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[29] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[18] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[18] * 100 / 6 = 33
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[19] * 100 / 6 = 33
```
```
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[2] * 100 / 6 = 100
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[3] * 100 / 6 = 100
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[4] * 100 / 6 = 100
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[5] * 100 / 6 = 100
(win32ss/user/ntuser/msgqueue.c:145) err: ajifRate[9] * 100 / 6 = 100
```
(printed from ReactOS)

Reply:
After adding the folllowing debug print:
`ERR("ajifRate[%d] * 100 / 6 = %d\n", ((PACON)NewCursor)->iicur, ((PACON)NewCursor)->ajifRate[((PACON)NewCursor)->iicur] * 100 / 6);`

Reply:
@katahiromz I just updated my PR by some new code. I found a better place for `IntKillTImer` call, and now it kills the timer and stops the animation when the current frame is not ACON (i. e., has no `CURSORF_ACONFRAME` flag). This is the best condition to top the animation, and it avoids unneeded endless animation even after the cursor was changed to another ani/non-ani (cur).